### ISSUE TITLE
How to add UNknown label for untrained images

### ISSUE THREAD

**Issue by guruvishnuvardan** [habrman/FaceRecognition#6]
HI,

I am using this code to identify unknown faces that are not trained, so Can you please let me know, where to add "Unknown" label, so that I can take screen shot of those unknown persons to train later.

Thanks
Guru

**Comment by habrman** [comment]
Hi,
I'm not sure that I understand what you want to do. The network is not trained on particular faces, instead it's trained to generate a vector given a face crop. This vector can be seen as a description of the face and as a point in a multidimensional space. If 2 of these points are close to each other they correspond to patches of the same person.

The way the code works now is that you match faces to the images in the `ids` folder. If it finds a match the name will be the folder name of the matching image and otherwise `unknown`. If you want a bunch of different people that have the id `untrained` you could just create a folder named `untrained` and place a bunch of images in there

**Comment by guruvishnuvardan** [comment]
HI,
Thanks a ton for a quick response. Yes, I have added the untrained folder, its working fine, but the system also recognizes a face, when there is no face, Can you please view the screen shot(top left, Dataset named Salmankhan is identified without a face). 
![screenshot from 2018-05-22 18-45-31](https://user-images.githubusercontent.com/31349523/40349229-8ff5770c-5dc3-11e8-8e03-d508c534ed62.png)


Can you please let me know, If I have to increase the training images, currently, I use 50 images

**Comment by habrman** [comment]
Can you enable the bounding box and landmark visualization? Can you also show me the images you have in the `Salmankhan` folder?

It's quite a noisy and unusual background. You clearly get a false positive face detection. The interesting thing is that you also get an id match. The distance of 0.7 is really small so I wonder if you have images that look like this in the `Salmankhan` folder. If you do, then you probably get a false positive face detection similar to the one in this image which would explain the match with the small distance.

**Comment by guruvishnuvardan** [comment]
Yes, I will try that and get back to you. Thanks.

**Comment by guruvishnuvardan** [comment]
HI,

Please note, these are the only images(in the name of Hrithik) provided to test the Last Samurai trailer, it works superbly in identifying faces as unknown. but Please find the screen shot, 
![screenshot from 2018-05-24 17-42-22](https://user-images.githubusercontent.com/31349523/40469255-83188674-5f4d-11e8-9529-5ea29a6a41be.png)



as the system identifies trees as unknown, I guess its a false negative scenario, Can you please let me know, how to avoid such scenarios.

![img27](https://user-images.githubusercontent.com/31349523/40469091-02cfd666-5f4d-11e8-8fd6-b762105c7a5a.jpg)

![img1](https://user-images.githubusercontent.com/31349523/40469081-011d8624-5f4d-11e8-96ba-3ecb6c033b40.jpg)
![img3](https://user-images.githubusercontent.com/31349523/40469082-015088bc-5f4d-11e8-877e-1fc145fa4fc5.jpg)
![img4](https://user-images.githubusercontent.com/31349523/40469083-01816e64-5f4d-11e8-8488-ed717127e413.jpg)
![img5](https://user-images.githubusercontent.com/31349523/40469085-01b3abea-5f4d-11e8-9fd4-d8cffe4b74c8.jpg)
![img8](https://user-images.githubusercontent.com/31349523/40469086-01e53d90-5f4d-11e8-996a-0d044248671c.jpg)
![img15](https://user-images.githubusercontent.com/31349523/40469087-0215260e-5f4d-11e8-91fe-130fdb70a65c.jpg)
![img18](https://user-images.githubusercontent.com/31349523/40469088-024399a8-5f4d-11e8-96d8-c2dbd674e3d5.jpg)
![img23](https://user-images.githubusercontent.com/31349523/40469089-0272b83c-5f4d-11e8-872a-764d0bbd1c72.jpg)
![img25](https://user-images.githubusercontent.com/31349523/40469090-02a0d4a6-5f4d-11e8-825e-629ca14f412c.jpg)

**Comment by habrman** [comment]
So this is not the same use-case as you posted before but the problem with false positives are still the same. You're getting false positive face detections with the cascade network and you can fix this by increasing the thresholds for the detector at this line:
https://github.com/habrman/FaceRecognition/blob/f1f20464d53266396202c16289f7aef12861d26b/detect_and_align.py#L442
There are three of them and if you read the paper you understand why. If you increase them the detector will have a more difficult time finding faces. You'll need to find a balance between finding real faces while minimizing the number of false positives. It might be enough to change only one of them, you'll simply have to play around.

**Comment by guruvishnuvardan** [comment]
Yes, I its working fine, not detecting false positives and face detection is also reduced in some scenarios, but thats OK.

Also, I face another problem, I connected an IP Camera to test the faces real time (http://en.uniview.com/res/201802/02/20180202_1611710_UNV%20IPC2122SR3-PF40(60)-C%202MP%20Mini%20Fixed%20Bullet%20Network%20Camera%20V2.3_789983_168459_0.pdf), but the frames per second is very slow, 2 to 3 FPS only, Can you please let me know, how to optimize the performance.

I use Ubuntu 16.04, CUDA 9.2, cuDNN 7.1, Tensorflow 1.8, NIVIDA GEFORCE GTX 970(4 GB RAM) GPU, AMD Opteron(tm) Processor 6320 × 16 processor.

**Comment by habrman** [comment]
Well the thing is that the networks are quite large so they take time. To speed thing up you could use [Haar Cascades from OpenCV](https://docs.opencv.org/3.4.1/d7/d8b/tutorial_py_face_detection.html) to find faces in the image. It's not as good but it should be faster than the network cascade I'm using now. You can't, however, get rid of the Inception Resnet which is a big network. Without thinking to much about it 3 possible solutions come to mind:

1. If you don't need to run on all frames I would recommend just skipping some of them. If you really need the facial position and ID in all frames you could also add a tracker instead of re-identifying all faces in each frame. This is the simplest solution. 

2. Another way would be to run multiple instances of the network. If you have 4 instances of the network you could potentially increase the fps by a factor of 4 but you would still have a latency.

3. Train your own network using a smaller architecture. As I mention in the README you can find the training code in this [repo](https://github.com/davidsandberg/facenet).

**Comment by guruvishnuvardan** [comment]
Hi,
Can you please clarify one of my questions, I have tried to identify my family persons using CCTV camera(2 Mega Pixels), but the system identifies only if the persons face is too close or sometimes it identifies as Unknown, as wall mounted CCTV camera is above 6 feet from the ground, do i need to train the faces from the camera's angle?

**Comment by habrman** [comment]
Your images are probably quite different from the dataset. Creating your own and training on it will defiantly help. You could also increase the distance threshold required for matching. It will lead to more false positive matches but it might be ok for your usecase
